src/utils.c

```diff
@@ -111,6 +111,16 @@ unmap_file(char *str, size_t size)
 
 #endif
 
+bool
+is_regular_file(const char *path)
+{
+    struct stat stat_buf;
+    int err = stat(path, &stat_buf);
+    if (err != 0) return false;
+
+    return S_ISREG(stat_buf.st_mode);
+}
+
```

Comment:
I'd argue that a simpler approach here would be
```c
FILE *open_regular_file(const char *path)
{
    if (!path) return NULL;
    ...
}
```

and in the caller you can then compress the code a bit too since you can do

```c
path = get_xcomposefile_path(ctx);
file = open_regular_file(path);
free(path);
if (file)
   goto out;

...
```

But the real benefit is that it helps removing a race condition having the `open` and the `stat` in the same function - you can/should do this:
```c
FILE *f = NULL;
int fd = open(path, O_RDONLY);
int err = fstat(fd, &stat_buf);
if (!S_ISREG(stat_buf.st_mode)) {
   fclose(fd);
   return NULL;
} 
file = fdopen(fd, "rb");
return file;
```

This way you don't have a race between checking if something is a directory and opening it that allows a potential attacker to swap out the file.

Reply:
A sane design, thank you!